kyo-core/shared/src/main/scala/kyo/scheduler/IOPromise.scala

```diff
@@ -263,80 +264,88 @@ private[kyo] object IOPromise:
     abstract class StateHandle:
         def compareAndSet[E, A](promise: IOPromise[E, A], curr: State[E, A], next: State[E, A]): Boolean
 
-    type State[+E, +A] = Result[E, A] | Pending[E, A] | Linked[E, A]
+    type State[E, A] = Result[E, A] | Pending[E, A] | Linked[E, A]
+
+    final case class Linked[E, A](p: IOPromise[E, A])
 
-    final case class Linked[+E, +A](p: IOPromise[E, A])
+    private val interruptPanic = Result.Panic(Interrupted(Frame.internal))
 
-    sealed abstract class Pending[+E, +A]:
+    sealed abstract class Pending[E, A]:
         self =>
 
         def waiters: Int
-        def interrupt[E2 >: E](v: Error[E2]): Pending[E, A]
+        def interrupt(v: Error[E]): Pending[E, A]
         def removeInterrupt(other: IOPromise[?, ?]): Pending[E, A]
-        def run[E2 >: E, A2 >: A](v: Result[E2, A2]): Pending[E2, A2]
+        def run(v: Result[E, A]): Pending[E, A]
 
         final def onComplete(f: Result[E, A] => Any): Pending[E, A] =
             new Pending[E, A]:
                 def waiters: Int = self.waiters + 1
-                def interrupt[E2 >: E](error: Error[E2]) =
-                    eval(discard(f(error.asInstanceOf[Error[E]])))
+                def interrupt(error: Error[E]) =
+                    eval(discard(f(error)))
                     self
                 def removeInterrupt(other: IOPromise[?, ?]) =
                     self.removeInterrupt(other).onComplete(f)
-                def run[E2 >: E, A2 >: A](v: Result[E2, A2]) =
+                def run(v: Result[E, A]) =
                     eval(discard(f(v.asInstanceOf[Result[E, A]])))
                     self
                 end run
 
         final def interrupts(p: IOPromise[?, ?]): Pending[E, A] =
             new Pending[E, A]:
-                def interrupt[E2 >: E](error: Error[E2]) =
-                    discard(p.interrupt(error.asInstanceOf[Error[E]]))
```

Comment:
This code was incorrect. If the current promise is interrupted with `Result.Failure[E]`, it'd be propagated to the `p` promise without a guarantee that it actually has `Failure[E]` in its error channel, which could cause a class cast exception. As an alternative, I'm keeping the panic if possible but using a placeholder in case it's an actual `Failure`.